### PR TITLE
refactor: Refactor brushes

### LINKED ISSUE #12601: WASM: Binding to a GradientStop Color on a LinearGradientBrush does not display the colour.
### Current behavior

No colour is displayed

This is the type of XAML.

                        <GradientStop Color="{x:Bind VM.ColourIndex, Mode=OneWay, Converter={StaticResource bodyColourConverter}, ConverterParameter=B}" Offset="0.5"/>

See the sample for the full context

### Expected behavior

Should work like the other platforms where the Color property can be bound

### How to reproduce it (as minimally and precisely as possible)

Project to reproduce is here:
https://github.com/BrianDT/UnoBoundColourGradientIssue.git

### Workaround

Hard code the color and have multiple shapes with a converter on the visibility

### Works on UWP/WinUI

Yes

### Environment

Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia

### NuGet package version(s)

Uno.WinUI 4.9.20

### Affected platforms

WebAssembly

### IDE

Visual Studio 2022

### IDE version

17.6.2

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #12601. Partially addresses #8891

## PR Type

What kind of change does this PR introduce?

### <samp>🤖 Generated by Copilot at 16c26cf</samp>

This pull request refactors various controls and classes to use the weak event proxy pattern for subscribing to brush changes, which improves memory management and performance. It also adds a new test class for the GradientBrush feature and a new base class for implementing the weak event proxy pattern. It deletes two unused files and fixes some minor issues with image brushes.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_GradientBrush.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+		var gradientStop2 = new GradientStop() { Color = Colors.Yellow, Offset = 1 };
+		rect.Fill = new LinearGradientBrush()
+		{
+			GradientStops = { gradientStop1, gradientStop2 }
+		};
+
+		WindowHelper.WindowContent = rect;
+		await WindowHelper.WaitForLoaded(rect);
+
+		gradientStop1.Color = Colors.Blue;
```
Comment (Youssef1313) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_GradientBrush.cs:28]: TODO: Assert that this change takes effect.

File: src/Uno.UI/UI/Xaml/Media/Brush.cs [hunk 2]
```diff
@@ -38,6 +83,7 @@ protected virtual void OnOpacityChanged(double oldValue, double newValue)
 
 		#endregion
 
+		// TODO: InvalidateRender when Transform/RelativeTransform changes, or their inner properties.
```
Comment (MartinZikmund) [inline src/Uno.UI/UI/Xaml/Media/Brush.cs:86]: Please create an issue for this TODO

**Review by MartinZikmund — APPROVED** [review]

**Comment by MartinZikmund** [comment]
Looks good to me!

**Comment by MartinZikmund** [comment]
Only thing I just realized now - could we introduce some race condition in case the brush is unsubscribed during GC / Java finalizer? As we could be removing handler from the event at the same time as adding another one on the UI thread, theoretically? @jeromelaban

File: src/Uno.UI/UI/Xaml/Media/Brush.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Media/Brush.cs:86]: This TODO is actually already done above, I think.

But now revising this part of the code again, it seems that the brush will now have a reference to the transforms, and I'm not sure the transforms can leak. We need to verify that very well.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Media/Brush.cs:86]: Double thinking, I think no leak concerns here. If a given transform is the current value of `Transform`/`RelativeTransform`, then that is owned by the brush anyway, and if the value changed, we unsubscribe from the event.

So it should be good I think.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_GradientBrush.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+using Windows.UI.Xaml.Media.Imaging;
+using Windows.UI.Xaml.Shapes;
+using static Private.Infrastructure.TestServices;
+
+namespace Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml_Media;
+
+[TestClass]
+public class Given_GradientBrush
+{
+	[TestMethod]
```
Comment (Youssef1313) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Media/Given_GradientBrush.cs:16]: TODO: This test needs `[RunsOnUIThread]` attribute

**Comment by Youssef1313** [comment]
Test failure on iOS:

```
Color at (0,0) is not expected
expected: FF0000FF #FF0000FF
actual  : FF0533FF #FF0533FF
tolerance: 5
difference: 51
```

The actual color is still bluish, meaning the GradientStop change propagated correctly. But I don't know why iOS diverges so much from blue. Anyway, that doesn't look like an issue with this PR, so I'll adjust tolerance for iOS.

File: src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs [hunk 4]
```diff
@@ -343,9 +356,12 @@ private static IDisposable InnerCreateLayer(UIElement owner, CALayer parent, Lay
...
 					acrylicBrush.Subscribe(owner, area, backgroundArea, parent, sublayers, ref insertionIndex, fillMask: null);
 				}
-				else if (background is XamlCompositionBrushBase unsupportedCompositionBrush)
+				else if (background is XamlCompositionBrushBase)
 				{
-					Brush.AssignAndObserveBrush(unsupportedCompositionBrush, color => backgroundLayer.FillColor = color)
+					Action onInvalidateRender = () => backgroundLayer.FillColor = Brush.GetFallbackColor(background);
+					background.InvalidateRender += onInvalidateRender;
+					onInvalidateRender();
+					new DisposableAction(() => background.InvalidateRender -= onInvalidateRender)
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs:364]: @Youssef1313 `disposables.Add(() => background.InvalidateRender -= onInvalidateRender)` should be available available to use.

File: src/Uno.UI/UI/Xaml/Controls/Panel/Panel.Android.cs [hunk 5]
```diff
@@ -20,15 +21,31 @@ namespace Windows.UI.Xaml.Controls
...
 		public Panel()
 		{
 			Initialize();
 		}
 
+		~Panel()
+		{
+			_backgroundBrushChangedProxy?.Unsubscribe();
+			_borderBrushChangedProxy?.Unsubscribe();
+		}
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Panel/Panel.Android.cs:40]: @Youssef1313 I don't think this is necessary. The proxy itself should be collected and that should be enough for the unsubscription to happen.

File: src/Uno.UI/UI/Xaml/Controls/Panel/Panel.Android.cs [hunk 6]
```diff
@@ -20,15 +21,31 @@ namespace Windows.UI.Xaml.Controls
...
+
+		protected override void JavaFinalize()
+		{
+			_backgroundBrushChangedProxy?.Unsubscribe();
+			_borderBrushChangedProxy?.Unsubscribe();
+			base.JavaFinalize();
+		}
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Panel/Panel.Android.cs:47]: Same here, if javafinalize is called, the .NET GC should be invoked soon after.

File: src/Uno.UI/UI/Xaml/Controls/Panel/Panel.Android.cs (hunk 5 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/Panel/Panel.Android.cs:40]: The brushes hold a reference to the proxy via the event subscription. So long living brushes will cause proxies to leak unless the control calls Unsubscribe.

File: src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs (hunk 4 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.iOSmacOS.cs:364]: @jeromelaban It should be the same functionality? or is there any difference?

Martin will be refactoring BorderLayerRenderer anyway.

**Pull request merged**